Tools: Implement better scrimmage support

araneg opened the pull request:
This change provides better scrimmage support by allowing a single scrimmage process to connect to multiple ArduPilot processes.

- sim_vehicle.py can now spawn multiple vehicles at once (prior behavior
  is maintained if the user decides not to use this feature)
- Users can specify multiple instances to spawn through either the -n or -i
  flags. Each instance executes in its own directory (<use_dir>/<instance_id>)
- The responsibility of spawning scrimmage has now been delegated to
  sim_vehicle.py instead of libraries/SITL/SIM_Scrimmage.cpp
- The mission.xml file which scrimmage expects is generated from a template,
  wherein each ardupilot instance corresponds to a scrimmage entity.
- All together, this allows for a single scrimmage simulator to connect to
  multiple instances of ardupilot, which in turn allows multiple ardupilot
  entities to interact with one another in scrimmage. Prior to this change,
  each ardupilot process was connected to its own scrimmage process.
- Misc. changes:
  * find_location_by_name() and find_new_spawn() have been rewritten such that a
    base location and coordinate offsets (specified by swarminit.txt) are
    maintained separately. This is necessary because scrimmage expects these
    values separately ({latitude,longitude,altitude}_origin vs {x,y,z})
  * find_autotest_dir() and find_root_dir() are now calculated only once. Issues
    were encountered when calling find_autotest_dir() after calling os.chdir()

Also, this change addresses some of the concerns presented in: https://github.com/ArduPilot/ardupilot/pull/11887

khancyr commented on Tools/autotest/sim_vehicle.py:
```diff
@@ -911,7 +939,7 @@ def generate_frame_help():
 group_sim.add_option("", "--use-dir",
                      type='string',
-                     default=None,
+                     default='.',
```
I am on phone and that change looks wrong. I will check later but it seems to launch sitl instance with root directory the one's where sim_vehicle.py is launched.

araneg replied:
. refers to the current working directory. I think this is consistent with the prior behavior. I made this change so that base_dir would always be set, even if the user did not specify --use-dir. This simplifies some of the code in start_vehicle and start_mavproxy. See: https://github.com/ArduPilot/ardupilot/pull/13405/files/b2ae81f729e513541d9f1d6bf84e1e0f324a9851#diff-460241a402d32677bcde6618e6e880deL1100-R1174

khancyr commented on Tools/autotest/sim_vehicle.py:
```diff
@@ -1137,10 +1205,42 @@ def generate_frame_help():
-                  loc=location)
+                  spawns=spawns)
+
+if cmd_opts.delay_start:
```
Duplicate from line 1245?

araneg replied:
I figured there should be a delay between starting scrimmage and mavproxy as well as between ardupilot and scrimmage.

khancyr commented on Tools/autotest/sim_vehicle.py:
```diff
@@ -1031,6 +1059,18 @@ def generate_frame_help():
+    print("May not specify a count less than 1")
+    sys.exit(1)
+
+if cmd_opts.scrimmage and (cmd_opts.vehicle != 'ArduPlane' and cmd_opts.vehicle != 'ArduCopter'):
```
Not in ["arduplane","arducopter"] maybe?

araneg replied:
Ok

khancyr commented on Tools/autotest/sim_vehicle.py:
```diff
@@ -1067,56 +1107,84 @@ def generate_frame_help():
     cmd_opts.frame = vinfo.options[cmd_opts.vehicle]["default_frame"]
 
-# setup ports for this instance
-mavlink_port = "tcp:127.0.0.1:" + str(5760 + 10 * cmd_opts.instance)
```
Were those unused?

araneg replied:
I factored this line (and the one after) into start_mavproxy
https://github.com/ArduPilot/ardupilot/pull/13405/files/b2ae81f729e513541d9f1d6bf84e1e0f324a9851#diff-460241a402d32677bcde6618e6e880deR749-R751

araneg commented:
> It is quite hard to review all, next time could you split the commit into group with logical changes? It really simplify the review process and we can still squash the commits later

Apologies for that. I'll see if I can separate the commit into multiple ones.

araneg commented:
Alright, I've split the commits, and fixed a few things. I've removed the --scrimmage option and now spawn scrimmage based on if the frame is set to either scrimmage-plane or scrimmage-copter. I've also addressed https://github.com/ArduPilot/ardupilot/pull/13405#discussion_r372122025
See: https://github.com/ArduPilot/ardupilot/pull/13405/commits/aad3b48926934f9d0b24777e62b53016fee011d7#diff-460241a402d32677bcde6618e6e880deL1089-R1168

peterbarker commented:
On Tue, 28 Jan 2020, araneg wrote:

>  *  All together, this allows for a single scrimmage simulator to connect to
>     multiple instances of ardupilot, which in turn allows multiple ardupilot
>     entities to interact with one another in scrimmage. Prior to this change,
>     each ardupilot process was connected to its own scrimmage process.

That's good... but others are using things like 
libraries/SITL/examples/Airsim/follow-copter.sh to achieve the same sort 
of thing.

Would this system be able to tidy all that up somehow?

Peter

araneg commented:
> That's good... but others are using things like libraries/SITL/examples/Airsim/follow-copter.sh to achieve the same sort of thing. Would this system be able to tidy all that up somehow? Peter

Perhaps. sim_vehicle currently can't spawn each ArduPilot process with different parameter files (as follow-copter.sh is doing), but with these changes adding that feature to sim_vehicle should be pretty simple.

Currently, scrimmage can only take mission parameters when it starts up. This, coupled with the need to pass scrimmage location and offset parameters lead me to modifying sim_vehicle as opposed to making a script. If I hadn't done that, I would have had to duplicate the code required for parsing locations.txt and swarminit.txt somewhere else. Plus, I think it's cleaner to be able to start multiple SITL simulations directly from sim_vehicle instead of a wrapper script.

peterbarker commented on Tools/autotest/sim_vehicle.py:
```diff
@@ -1155,5 +1254,7 @@ def generate_frame_help():
         start_mavproxy(cmd_opts, frame_infos)
 except KeyboardInterrupt:
     progress("Keyboard Interrupt received ...")
+    if tmp is not None:
```
Maybe use `atexit`?

tridge commented:
needs testing with non-scrimmage to ensure the changed options (such as mcast) still work

araneg commented:
Logs: http://0x0.st/iXHQ.gz
Tests:
sim_vehicle.py -vArduPlane -fplane-jet -LElliottField -I101 -m --map --console
sim_vehicle.py -vArduPlane -fplane -LElliottField --mcast -I101 -m --map --console
sim_vehicle.py -vArduPlane -fplane -LElliottField -i'1 2' -Z swarminit.txt -m --map --console
sim_vehicle.py -vArduPlane -fplane -LElliottField --mcast -i'1 2' -Z swarminit.txt -m --map --console

The first three ran fine, but the last one caused the plane in the mavproxy map to rapidly switch between the locations of the two instances. That being said, I tried running two separate instances of sim_vehicle with multicast like so:
sim_vehicle.py -vArduPlane -fplane -LElliottField --mcast -I1 -Z swarminit.txt -m --map --console
sim_vehicle.py -vArduPlane -fplane -LElliottField --mcast -I2 -Z swarminit.txt -m --map --console
And got the same results. I also tested against master with the same results. I figure this is to be expected, since MAVProxy is listening on multicast, it probably can't differentiate between the two arduplanes.

khancyr commented on Tools/autotest/template/ardupilot.xml:
```diff
@@ -0,0 +1,101 @@
+<?xml version="1.0"?>
```
shouldn't this be template_scrimmage ?

araneg replied:
Ah, yes. Good catch.

khancyr commented:
I have tried :
`sim_vehicle.py -vArduCopter -f + -I2 --use-dir plop --map --console`
It works as intended.

`sim_vehicle.py -vArduCopter -f + -I2 --use-dir plop -n2 --map --console`
It does create two directories under plop, with instance number.
The I parameter is overrides by the count one, and ports number are well scaled.
Two mavproxy instances are launched.

One issue I find is that the console and map from the second instance aren't closed on exit.

same behavior with `sim_vehicle.py -vArduCopter -f + -I2 --use-dir plop -n3 --map --console`

`sim_vehicle.py -vArduCopter -f + -I2 --use-dir plop -n2 --mcast --map --console`
add mcast is working and we get the two vehicles. Now we get back on the issue of not have the sysid set correctly but that another story.

Looks nice improvement for multivehicle.

peterbarker commented on Tools/autotest/sim_vehicle.py:
```diff
@@ -231,6 +231,7 @@ def kill_tasks():
             'AntennaTracker.elf',
             'JSBSIm.exe',
             'MAVProxy.exe',
+            'mavproxy.py',
```
This is kind of nasty; it's going to kill every mavproxy on the system, on all platforms.  I'm often running several MAVProxys and I'd prefer them not all be killed when we run through here.

I do acknowledge it seems to have this behaviour under Windows :-)

khancyr replied:
And it doesn't fix the issue of not closing the map and console windows !

araneg replied:
@peterbarker
I agree that this is nasty, but it's basically how tasks were killed before. I think an apt solution is to populate a global set with PIDs of each process as they are spawned. Upon exit, that set would be used to kill processes as opposed to a set of names.

@khancyr
Strange, it works on my end. Are you on Linux?

khancyr replied:
yep, I am Ubuntu 19.10. that doesn't kill all mavproxy process

khancyr replied:
ok I find the error, I don't have psutil ...

khancyr replied:
About why the remain windows aren't closed, it seems that the run_in_terminal command is creating a fork for the mavproxy instance.
That is annoying as we don't have a way to get the pid to kill them

peterbarker commented:
On Mon, 24 Feb 2020, araneg wrote:

> I agree that this is nasty, but it's basically how tasks were killed before. I think an apt solution is to populate a global set with PIDs of each process
> as they are spawned. Upon exit, that set would be used to kill processes as opposed to a set of names.

Sadly that doesn't work as any process which spawns more processes may 
leave those (grandchild processes) dangling (MAVProxy....)

Best I've ever come up with in this situation is to stain the environment 
of the process with a specific name/value pair.  When you want to kill a 
process you go through the entire list of processes and kill off anything 
which has that name/value pair in the environment.

... also, sorry, but we *don't* currently do this on Linux, and it would 
not be well-received by a lot of devs.  Windows.... well...

Perhaps in the specific mavproxy case we could jot down the PID and kill 
it with an atexit?  Testing required for that one!

araneg commented:
> Sadly that doesn't work as any process which spawns more processes may leave those (grandchild processes) dangling (MAVProxy....)

Good point. Maybe we could recurse over the children of each process spawned by sim_vehicle.py. psutil provides a children() method. Of course, that wouldn't work in the case that the process double forks and reparents onto PID 1, but it doesn't seem like that's the case here. This is what pstree shows (after modifying run_in_terminal_window.sh to use exec and removing &):
```
python─┬─mavproxy.py───2*[mavproxy.py]
       ├─2*[xterm───arduplane]
       └─xterm───mavproxy.py───2*[mavproxy.py]
```
That's when running two SITLS. I didn't see any arduplanes or mavproxys outside of this tree.
To avoid reparenting as we're killing process, we could either kill processes depth first, or just populate a set of PIDs which includes all children, and then kill based on that set. I don't know about the case where psutil isn't present though.

Another solution (which is unportable) is to setup sim_vehicle.py as a subreaper so all children reparent onto sim_vehicle, but this requires executing the prctl sys-call, which requires ctypes which becomes messy and overkill.

> ... also, sorry, but we *don't* currently do this on Linux, and it would
not be well-received by a lot of devs.  Windows.... well...

Maybe I'm missing something, but it seems like kill_tasks_psutil simply iterates over all processes and kills a process on a name match. kill_tasks_pkill effectively does the same thing. Are you setting the argv0 of the spawned processes to a custom value somewhere?
Unless by "this" you meant specifically killing mavproxy.py? In which case, I agree.

khancyr commented:
hee, your modification is giving a good tree ! 
Instead of killing we should send terminate signal as it will lead all children to terminate too ! And kill what remain. There is and example in psutil documentation on how to do it. 
I will add psutil in the install list, it will really help

araneg commented:
> Instead of killing we should send terminate signal as it will lead all children to terminate too !
I don't think SIGTERM kills children. At least not generally/by default.
> There is and example in psutil documentation on how to do it.
Are you talking about this: https://psutil.readthedocs.io/en/latest/#kill-process-tree

araneg commented:
I got it working, but it depends on psutil which doesn't support cygwin. It'll take some more refinement.
I had to determine children prior to a keyboard interrupt being received because upon Ctrl-C mavproxy is killed which ruins the process tree. The map and console of any mavproxy process running via run_in_terminal_window.sh become reparented unfortunately.

peterbarker commented:
On Tue, 25 Feb 2020, araneg wrote:

> Unless by "this" you meant specifically killing mavproxy.py? In which case, I agree.

Yes, mavproxy.py specifically.

khancyr commented on Tools/autotest/sim_vehicle.py:
```diff
@@ -1250,7 +1242,7 @@ def generate_frame_help():
         time.sleep(3)  # output our message after run_in_terminal_window.sh's
         progress("Waiting for SITL to exit")
-        wait_unlimited()
+        signal.pause()
```
we cannot use that as windows don't support it

araneg replied:
Okay, I'll revert that change.

khancyr commented:
hello,

The last push broken everything. It fails now to launch SITL even for 1 vehicle, and there is still issue with a map window. I guess we should have Mavproxy cleaning up its process cleanly before exiting.

khancyr commented on Tools/autotest/run_in_terminal_window.sh:
```diff
@@ -31,17 +31,17 @@ if [ -n "$SITL_RITW_TERMINAL" ]; then
       ICONIC=-iconic
   fi
-  exec xterm $ICONIC -xrm 'XTerm*selectToClipboard: true' -xrm 'XTerm*initialFont: 6' -n "$name" -name "$name" -T "$name" -hold -e $*
+  exec xterm "$ICONIC" -xrm 'XTerm*selectToClipboard: true' -xrm 'XTerm*initialFont: 6' -n "$name" -name "$name" -T "$name" -hold -e "$@"
```
double quote a breaking the cmdline

araneg replied:
My apologies. Fixed.

khancyr commented on Tools/autotest/run_in_terminal_window.sh:
```diff
@@ -34,7 +34,7 @@ elif [ -n "$DISPLAY" -a -n "$(which xterm)" ]; then
   if [ "$SITL_RITW_MINIMIZE" -eq 1 ]; then
       ICONIC=-iconic
   fi
-  exec xterm "$ICONIC" -xrm 'XTerm*selectToClipboard: true' -xrm 'XTerm*initialFont: 6' -n "$name" -name "$name" -T "$name" -hold -e "$@"
```
wrong fix !;-P

araneg replied:
"$@" should be fine, it retains arguments as seperate. Are you setting --mavproxy-args in sim_vehicle? I thought line 660-691 in sim_vehicle would do word splitting.

mday299 commented:
bump

araneg commented:
>I guess we should have Mavproxy cleaning up its process cleanly before exiting.

I've opened a pull request on the mavproxy repo about this.

See: https://github.com/ArduPilot/MAVProxy/pull/739

araneg commented:
What's the status on this?

khancyr commented on Tools/autotest/sim_vehicle.py:
```diff
@@ -31,6 +31,8 @@
 autotest_dir = os.path.dirname(os.path.realpath(__file__))
 root_dir = os.path.realpath(os.path.join(autotest_dir, '../..'))
 
+os.environ["SIM_VEHICLE_SESSION"] = os.urandom(8).encode('hex')
```
binascii.hexlify(os.urandom(8)).decode() to work with  python3

peterbarker commented:
Just got this:

```
SIM_VEHICLE: kill_tasks failed: 'Process' object has no attribute 'environ'
```

araneg commented:
It looks like psutil isn't exposing the environment on your system. .environ should be cross platform.
https://psutil.readthedocs.io/en/latest/#psutil.Process.environ
Which version of psutil do you have?

araneg commented:
I've made a commit which falls back onto using process names for killing tasks if psutil doesn't expose process environments.

https://github.com/araneg/ardupilot/commit/e048f6ca5505a88744f5411dbcd4217590422316

If your version of psutil is already newer than 4.0.0, then can you test this commit and let me know if it works? I'll open another PR afterwards.

peterbarker commented:
```
SIM_VEHICLE: Keyboard Interrupt received ...
SIM_VEHICLE: Killing tasks
SIM_VEHICLE: kill_tasks failed: 'Process' object has no attribute 'environ'
pbarker@bluebottle:~/rc/ardupilot(master)$ python
Python 2.7.17 (default, Apr 15 2020, 17:20:14) 
[GCC 7.5.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import psutil
>>> psutil.__version__
'3.4.2'
>>> 
```

peterbarker commented:
That version's *probably* too old to carry around sepcialised code for.

It's probably sufficient to add a dependency to `requirements.txt` to ensure a recent-enough version of `psutil`?  What's the oldest version we can specify which will still have the functionality required?

araneg commented:
.environ was added in psutil 4.0.0.
Alternatively, can you pull https://github.com/araneg/ardupilot/commit/e048f6ca5505a88744f5411dbcd4217590422316 and to see if the problem persists.

araneg commented:
Nevermind, I just saw your comment.

khancyr commented:
@peterbarker we don't have requirement.txt for SITL/sim_vehicle.py ... but could create one, or add a specific version to the prereq script

peterbarker commented:
On Wed, 29 Apr 2020, Pierre Kancir wrote:

> @peterbarker we don't have requirement.txt for SITL/sim_vehicle.py ... but could create one, or add a specific version to the prereq script

Or an explicit check in sim_vehicle.py?

*and*, perhaps, a requirements.txt - which we could use as part of install 
but also give someone a file to point pip at....
